lib/Semantic_check.ml

```diff
@@ -573,153 +594,45 @@ and semantic_check_expression cf {expr_untyped_loc= loc; expr_untyped} =
-              let _ =
-                if
-                  not
-                    (check_compatible_arguments_mod_conv uid.name listedtypes
-                       ues)
-                then
-                  semantic_error ~loc
-                    ( "Ill-typed arguments supplied to function "
-                    ^ ("'" ^ uid.name ^ "'")
-                    ^ ". Available signatures:\n"
-                    ^ pretty_print_unsizedtype
-                        (UFun (listedtypes, ReturnType ut))
-                    ^ "\nInstead supplied arguments of incompatible type: "
-                    ^ pretty_print_unsizedtypes
-                        (List.map ~f:type_of_expr_typed ues)
-                    ^ "." )
-              in
-              { expr_typed= FunApp (uid, ues)
-              ; expr_typed_ad_level= returnblock
-              ; expr_typed_type= ut
-              ; expr_typed_loc= loc }
-          | Some _ ->
-              (* Check that Funaps are actually functions *)
-              semantic_error ~loc
-                ( "A returning function was expected but a non-function value "
-                ^ ("'" ^ uid.name ^ "'")
-                ^ " was supplied." )
-          | None ->
-              semantic_error ~loc
-                ( "A returning function was expected but an undeclared \
-                   identifier "
-                ^ ("'" ^ uid.name ^ "'")
-                ^ " was supplied." ) ) )
+            ^ pretty_print_unsizedtype ue.expr_typed_type
+            ^ "." ) )
+  | Variable id ->
+      let uid = semantic_check_identifier id in
+      let ut = Symbol_table.look vm id.name in
+      (* Check that variable in scope if used  *)
+      let _ =
+        if
+          ut = None
+          && not (Stan_math_signatures.is_stan_math_function_name uid.name)
+        then
+          semantic_error ~loc
+            ("Identifier " ^ ("'" ^ uid.name ^ "'") ^ " not in scope.")
+      and originblock, type_ =
+        Option.value ~default:(MathLibrary, Mir.UMathLibraryFunction) ut
+      in
+      { expr_typed= Variable uid
+      ; expr_typed_ad_level= calculate_autodifftype originblock type_
+      ; expr_typed_type= type_
+      ; expr_typed_loc= loc }
+  | IntNumeral s ->
+      { expr_typed= IntNumeral s
+      ; expr_typed_ad_level= DataOnly
+      ; expr_typed_type= UInt
+      ; expr_typed_loc= loc }
+  | RealNumeral s ->
+      { expr_typed= RealNumeral s
+      ; expr_typed_ad_level= DataOnly
+      ; expr_typed_type= UReal
+      ; expr_typed_loc= loc }
+  | FunApp (_, id, es) ->
+      let uid = semantic_check_identifier id
+      and ues = List.map ~f:(semantic_check_expression cf) es in
+      semantic_check_fn_map_rect ~loc (uid, ues)
+      |> Or_error.bind ~f:(semantic_check_fn_conditioning ~loc)
+      |> Or_error.bind ~f:(semantic_check_fn_target_plus_equals cf ~loc)
+      |> Or_error.bind ~f:(semantic_check_fn_rng cf ~loc)
+      |> Or_error.bind ~f:(semantic_check_fn ~loc)
```

Comment:
seems like we could use `val Or_error.combine_errors_unit : unit Or_error.t list -> unit Or_error.t` maybe?

Reply:
There is definitely no need to `bind` the first 4 checks since they don't depend on one another. Since we aren't operating on the input in these functions, the return type of these _should_ be `unit Or_error`.

The last function `semantic_check_fn` is actually doing something and returning a `typed_expression` with the correct `fun_kind` so we can't use `Or_error.combine_errors_unit`.

What we want is to have the effectful functions composed to drop the `unit Or_error.t` but keep the result of the last function. 

There is an infix function `*>` from `Applicative` which does what we want - sequences actions and ignores the first argument. It's a bit messy to [write in terms of `apply`](https://github.com/janestreet/base/blob/master/src/applicative.ml#L50) so I think it's worth paying the price here.

The code would look like:

```
      Or_error.(
        semantic_check_fn_map_rect ~loc (uid, ues)
        *> semantic_check_fn_conditioning ~loc (uid, ues)
        *> semantic_check_fn_target_plus_equals cf ~loc (uid, ues)
        *> semantic_check_fn_rng cf ~loc (uid, ues)
        *> semantic_check_fn ~loc (uid, ues)
        |> ok_exn 
      )
```